### PR TITLE
Update TensorRT-LLM

### PR DESCRIPTION
**Pull request by kaiyux**
* Model Support
  - BART and mBART support in encoder-decoder models
  - Support FairSeq Neural Machine Translation (NMT) family
  - Mixtral-8x7B model support
    - Support weight loading for HuggingFace Mixtral model
* Features
  - MPT - Int4 AWQ / SmoothQuant support
  - Support speculative decoding with prefilled KV cache
  - Support AWQ and GPTQ support for QWEN
  - Support ReduceScatter plugin
* Bug fixes
  - Fix tokenizer usage in quantize.py #288, thanks to the contribution from @0xymoro
  - Fix LLaMa with LoRA error #637
  - Fix the first token being abnormal issue when '--gather_all_token_logits' is enabled #639
  - Fix Expert and Tensor Parallelism for Mixture of Expert layers
  - Fix an accuracy bug in MMHA `multi_block_mode` when certain conditions are met (large TP & 32K sequence length)
  - Fix BERT FMHA flag
* Performance
  - Optimize Hopper warp specialized kernels
  - Optimize AllReduce for parallel attention on Falcon and GPT-J
  - Enable split-k for weight-only cutlass kernel when SM>=75

### CODE REVIEW

**Review by Shixiaowei02 — APPROVED** [review]

File: cpp/tensorrt_llm/runtime/worldConfig.cpp [hunk 1]
```diff
@@ -65,17 +67,66 @@ void initMpi(nvinfer1::ILogger& logger, int threadMode = MPI_THREAD_FUNNELED)
...
+        {
+            TLLM_LOG_WARNING("The user specified device IDs are not contiguous!");
+        }
+        TLLM_LOG_INFO("Using user-specified devices: %s", tc::arr2str(mDeviceIds.data(), numDevices).c_str());
+    }
+
+    TLLM_CHECK(mTensorParallelism > 0);
+    TLLM_CHECK(mPipelineParallelism > 0);
+
+    TLLM_CHECK_WITH_INFO(static_cast<SizeType>(numDevices) >= tensorParallelism * pipelineParallelism,
```
Comment (leavelet) [inline cpp/tensorrt_llm/runtime/worldConfig.cpp:113]: There seems to be a mistake here? In a multi-node setup, the product of `mGpusPerNode` and the number of nodes should exceed the total of `TP` multiplied by `PP`. However, in this case, the number of GPUs per node is actually less than the product of `TP` and `PP`
@kaiyux

Comment (juney-nvidia) [reply cpp/tensorrt_llm/runtime/worldConfig.cpp:113]: Thanks for the sharp catch. I will discuss with the TensorRT-LLM engineers working on C++ runtime and go back to you later. 

June

Comment (leavelet) [reply cpp/tensorrt_llm/runtime/worldConfig.cpp:113]: Hello, is there any progress on this?

Comment (kaiyux) [reply cpp/tensorrt_llm/runtime/worldConfig.cpp:113]: Hi @leavelet , we did not fully test multi-node support in TensorRT-LLM, it might be working, but there is no guarantee. If you need to try that, we suggest that you remove the check locally so that it won't block you. Thanks very much for your interest on our work!
